graphbolt/include/graphbolt/rowwise_pick.h

```diff
@@ -0,0 +1,184 @@
+  const bool has_probs = probs.has_value();
+  auto thread_num = compute_num_threads(0, num_rows, kDefaultPickGrainSize);
+  std::vector<size_t> block_offset(thread_num + 1, 0);
+
+  torch::parallel_for(
+      0, num_rows, kDefaultPickGrainSize, [&](size_t b, size_t e) {
+        int64_t sampled_num_this_thread = 0;
+        for (size_t i = b; i < e; ++i) {
+          const auto rid = rows[i].item<int64_t>();
+          TORCH_CHECK(rid >= 0 && rid < graph->NumNodes());
+          const auto off = indptr[rid].item<int64_t>();
+          const auto len = indptr[rid + 1].item<int64_t>() - off;
+
+          if (len == 0) continue;
+
+          torch::Tensor picked_indices_this_row;
+          // Fast path.
+          if ((pick_all || len <= min_num_pick) && !replace) {
+            picked_indices_this_row =
+                torch::arange(off, off + len, indptr.options());
+            if (has_probs) {
+              auto mask_tensor = probs.value().slice(0, off, off + len) > 0;
+              picked_indices_this_row =
+                  torch::masked_select(picked_indices_this_row, mask_tensor);
+            }
+          } else {
+            TensorList pick_indices_per_etype(
+                num_etypes, torch::tensor({}, indptr.options()));
+            auto cur_et = type_per_edge_accessor[off];
+            auto cur_num_pick = num_picks_accessor[cur_et];
+            int64_t et_len = 1;
+            for (int64_t j = 0; j < len; ++j) {
+              TORCH_CHECK(
+                  j + 1 == len || cur_et <= type_per_edge_accessor[off + j + 1],
+                  "Edge type is not sorted. Please sort in advance");
+
+              if ((j + 1 == len) ||
+                  cur_et != type_per_edge_accessor[off + j + 1]) {
+                // 1 end of the current etype.
+                // 2 end of the row.
+                // Random pick for current etype.
+                if (cur_num_pick != 0) {
+                  torch::Tensor picked_indices;
+                  int64_t et_end = off + j + 1;
+                  int64_t et_start = et_end - et_len;
+                  if ((cur_num_pick == -1) ||
+                      (et_len <= cur_num_pick && !replace)) {
+                    // Fast path.
+                    picked_indices =
+                        torch::arange(et_start, et_end, indptr.options());
+                    if (has_probs) {
+                      auto mask_tensor =
+                          probs.value().slice(0, et_start, et_end) > 0;
+                      picked_indices =
+                          torch::masked_select(picked_indices, mask_tensor);
+                    }
+                  } else {
+                    picked_indices = pick_fn(et_start, et_end, cur_num_pick);
+                  }
+                  pick_indices_per_etype[cur_et] = picked_indices;
+                }
+
+                if (j + 1 == len) break;
+
+                cur_et = type_per_edge_accessor[off + j + 1];
+                et_len = 1;
```

Comment:
Please take a look at @czkkkkkk 's comment. The good practice is to avoid mutable loop variables as much as possible (unless they are in the loop expression).

Reply:
Changed